automation/JSInstrumentation.py

```diff
@@ -0,0 +1,204 @@
+import json
+import os
+
+import jsonschema
+
+curdir = os.path.dirname(os.path.realpath(__file__))
+schema_path = os.path.join(
+    curdir, 'js_instrumentation', 'js_instrument_modules.schema'
+)
+
+list_log_settings = [
+    'propertiesToInstrument',
+    'nonExistingPropertiesToInstrument',
+    'excludedProperties'
+]
+shortcut_specs = {
+    'fingerprinting':
+    os.path.join(curdir, 'js_instrumentation', 'fingerprinting.json')
+}
+
+
+def get_default_log_settings():
+    return {
+        'propertiesToInstrument': [],
+        'nonExistingPropertiesToInstrument': [],
+        'excludedProperties': [],
+        'logCallStack': False,
+        'logFunctionsAsStrings': False,
+        'logFunctionGets': False,
+        'preventSets': False,
+        'recursive': False,
+        'depth': 5,
+    }
+
+
+def python_to_js_string(py_in):
+    """Takes python in and converts it to a string
+    of the equivalent JS object.
+
+    Customized for our specific needs:
+    * expects a list
+    * object is de-quoted
+    """
+    objects = [x['object'] for x in py_in]
+    out = json.dumps(py_in)
+    for o in objects:
+        obj_str_before = f'"object": "{o}",'
+        obj_str_after = f'"object": {o},'
+        out = out.replace(obj_str_before, obj_str_after)
+    return out
+
+
+def validate(python_list_to_validate):
+    schema = json.loads(open(schema_path).read())
+    jsonschema.validate(instance=python_list_to_validate, schema=schema)
+    # Check properties to instrument and excluded properties don't collide
+    for request in python_list_to_validate:
+        propertiesToInstrument = \
+            request['logSettings']['propertiesToInstrument']
+        if propertiesToInstrument is not None:
+            propertiesToInstrument = set(propertiesToInstrument)
+            excludedProperties = set(
+                request['logSettings']['excludedProperties'])
+            assert len(propertiesToInstrument.intersection(
+                excludedProperties)) == 0
+    return True
+
+
+def merge_object_requests(python_list):
```

Comment:
Why support this at all? It seems safer to throw an error if there are duplicate configurations for the same object. Especially given that this isn't 100%.

Reply:
I guess maybe this might happen from recursive instrumentation?

Reply:
To make it easier to mix collections and API requests. I still think this is a good idea. We can remove it if it causes drama.